src/vision/status_quo/alan_builds_a_task_scheduler.md

```diff
@@ -0,0 +1,704 @@
+        }
+    }
+
+    pub fn register(&self, id: u64) -> io::Result<crossbeam_channel::Receiver<String>> {
+        let mut _mu;
+        match self.requests.try_write() {
+            Some(guard) => _mu = guard,
+            None => return Err(...),
+        }
+
+        let (request_tx, request_rx) = unbounded();
+        if _mu.get(&id).is_none() {
+            _mu.insert(id, request_tx);
+        } else {
+            return Err(...)
+        }
+
+        Ok(request_rx)
+    }
+
+    pub fn trigger(&self, id: u64, x: String) -> io::Result<()> {
+        let mut _mu;
+        match self.requests.try_write() {
+            Some(guard) => _mu = guard,
+            None => return Err(...),
+        }
+
+        let request_tx;
+        match _mu.get(&id) {
+            Some(ch) => request_tx = ch,
+            None => return Err(...),
+        }
+
+        match request_tx.send(x) {
+            Ok(_) => _mu.remove(&id),
+            Err(e) => return Err(...),
+        }
+
+        Ok(())
+    }
+}
+```
+
+Alan is still not sure if `crossbeam_channel` is safe for async programming and whether he should instead use another crate [`async_std::channel`](https://docs.rs/async-std/1.9.0/async_std/channel/index.html). While `crossbeam_channel` seems to work, Alan is not confident about his choice. Disgruntled with seemingly unnecessary divergence in the community, Alan wonders why all those cool improvements had not been made back to Rust core std libraries.
+
+### **Implementing task applier**
+
+Alan implements a task applier, which simply echoes the requested message, as in Go:
+
+```go
+type EchoManager interface {
+	apply(req *EchoRequest) (string, error)
+}
+
+type echoManager struct {
+	mu sync.RWMutex
+}
+
+func (ea *echoManager) apply(req *EchoRequest) (string, error) {
+	ea.mu.Lock()
+	defer ea.mu.Unlock()
+	switch req.Kind {
+	case "create":
+		return fmt.Sprintf("SUCCESS create %q", req.Message), nil
+	case "delete":
+		return fmt.Sprintf("SUCCESS delete %q", req.Message), nil
+	default:
+		return "", fmt.Errorf("unknown request %q", req)
+	}
+}
+```
+
+Having implemented event notifier above, Alan is now somewhat familiar with Rust mutex and writes the following Rust code:
+
+```rust
+// 1st version
+use async_std::sync::RwLock;
+
+pub struct Manager {
+    mu: RwLock<()>,
+}
+
+impl Manager {
+    pub fn new() -> Self {
+        Self {
+            mu: RwLock::new(()),
+        }
+    }
+
+    pub fn apply(&self, req: &Request) -> io::Result<String> {
+        let _mu;
+        match self.mu.try_write() {
+            Some(guard) => _mu = guard,
+            None => return Err(...),
+        }
+        match req.kind.as_str() {
+            "create" => Ok(format!(
+                "SUCCESS create {}",
+                to_string(req.message.to_owned())
+            )),
+            "delete" => Ok(format!(
+                "SUCCESS delete {}",
+                to_string(req.message.to_owned())
+            )),
```

Comment:
Stringly-typed API is not a good idea in rust, we should switch this to something else. Like `Create` and `Delete` within an enum.